scripts/benchmarks/benchmark.py

```diff
@@ -308,6 +316,38 @@ def run(self, run_cmd: str, environment_variables: Dict = None):
         self.environment = environment_variables
         self.experiment_args_str = commands
         os.makedirs(self.save_dir, exist_ok=True)
+
+        if log_memory:
+            '''
+            Opens a parallel process to log the device memory of the main experiment process. 
+            - Logs memory at intervals to a csv file in `self.save_dir`
+            - Terminates at the end of experiment
+            - GPU log is read and aggregated when the experiment ends & results are saved in Experiment.write_result, 
+
+            NOTE: assumes the experiment is the only process on the gpu devices -
+            there are no other processes running on the device in parallel.
+
+            Can log more details from nvidia-smi by expanding GPU_Table argument
+            e.g. "timestamp,name,index,utilization.gpu,utilization.memory,memory.total,memory.free,memory.used"
+            '''
+            assert torch.cuda.device_count()>0, "No device detected for memory logging!"
+            nvidia_logging_cmd = [
+                "nvidia-smi",
+                "--query-gpu",
+                GPU_TABLE,
+                "--format",
+                "csv",
+                "--id",
+                str(environment_variables['CUDA_VISIBLE_DEVICES']),
```

Comment:
we need to be careful of this
- outside we set `CUDA_VISIBLE_DEVICES` in environment_variables to 0,1...num_gpus
- then here we take them from the environment of `benchmark.py` which can be arbitrary, e.,g. `CUDA_VISIBLE_DEVICES=3,4,6,...`
- so we must be sure that doing this 3 -> 0, 4 -> 1, etc.
- and we must document this down if it is correct

when you print the CSV in the `gpu_log_fileanme`, will be gpu ids be printed as 3,4,6.. or 0,1,2?

also what if CUDA_VISIBLE_DEVICES is not set?

Reply:
Okay i amended the code for all experiments to reference their device ids from `benchmark.py`'s `CUDA_VISIBLE_DEVICES` environment variable and if the env var is not set, default to taking the device ids from an enumeration of `torch.cuda.device_count`